File: parser.ts
import puppeteer from 'puppeteer';
import fs from 'fs';
import path from 'path';

async function downloadJsonFile(url: string) {
    // Путь к текущей папке, где находится скрипт
    const downloadPath = path.join(__dirname);

    // Запускаем браузер с аргументами для скачивания в текущую директорию
    const browser = await puppeteer.launch({
        headless: true,
        args: [
            `--disable-web-security`,
            `--disable-features=IsolateOrigins,site-per-process`,
            `--no-sandbox`,
            `--disable-setuid-sandbox`,
            `--disable-dev-shm-usage`,
            `--disable-software-rasterizer`,
            `--disable-gpu`,
            `--window-size=1920x1080`,
            `--disable-extensions`,
            `--disable-infobars`,
            `--disable-translate`,
            `--download-dir=${downloadPath}` // Указываем путь для загрузок
        ],
        defaultViewport: null,
    });

    const page = await browser.newPage();

    // Переходим на страницу
    console.log(`Navigating to ${url}...`);
    await page.goto(url, { waitUntil: 'networkidle2' });

    // Получаем содержимое страницы
    const pageContent = await page.content();

    // Сохраняем содержимое страницы в файл
    const filePath = path.join(downloadPath, 'page_content.html');
    fs.writeFileSync(filePath, pageContent, 'utf8');
    console.log('Page content has been saved to:', filePath);

    // Селектор для кнопки
    const downloadButtonSelector = 'button.MuiButton-outlined';

    try {
        console.log('Waiting for button to become visible...');
        // Ожидаем, пока кнопка станет доступной с увеличенным таймаутом
        await page.waitForSelector(downloadButtonSelector, { timeout: 15000 }); // увеличиваем таймаут до 15 секунд
        console.log('Button found!');

        // Прокручиваем страницу, чтобы кнопка оказалась в видимой области
        console.log('Scrolling to the button...');
        await page.evaluate((selector) => {
            const button = document.querySelector(selector);
            if (button) {
                button.scrollIntoView({ behavior: 'smooth', block: 'center' });
            }
        }, downloadButtonSelector);

        console.log('Waiting for the button to be clickable...');
        // Пробуем кликнуть через JavaScript (evaluate)
        const button = await page.$(downloadButtonSelector);
        if (button) {
            // Используем evaluate для клика через JS, чтобы обойти ограничение Puppeteer
            await page.evaluate((btn) => {
                btn.click();
            }, button);
            console.log('Button clicked!');
        } else {
            console.error('Button is not found after scrolling');
        }

        // Ожидаем скачивания файла (проверка происходит по времени ожидания)
        console.log('Waiting for file download...');
        await page.waitForTimeout(3000); // Ждем 3 секунды для скачивания

        // Скачиваем файл в текущую директорию
        const downloadedFilePath = path.join(downloadPath, 'lobby_data.json');
        const data = await page.content();
        fs.writeFileSync(downloadedFilePath, data);

        console.log('JSON файл скачан в:', downloadedFilePath);
    } catch (error) {
        console.error('Error while waiting for button or clicking:', error);
    }

    await browser.close();
}

// Пример использования
const url = 'https://eternalesports.club/lobbies/3/19882b'; // Замени на актуальный URL

// Запускаем функцию
downloadJsonFile(url);

File: parser.py
import os
import time
import chromedriver_autoinstaller
from selenium import webdriver
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

# Устанавливаем нужную версию ChromeDriver, если ещё не установлена
chromedriver_autoinstaller.install()

# Определяем путь для скачивания
download_path = os.getcwd()  # Текущая рабочая директория, где находится скрипт

# Настроим параметры для Chrome
chrome_options = Options()
chrome_options.add_argument("--headless")  # Запуск в фоновом режиме
chrome_options.add_argument("--no-sandbox")
chrome_options.add_argument("--disable-dev-shm-usage")
chrome_options.add_argument("--disable-extensions")

# Указываем путь для скачивания с помощью Chrome Preferences
prefs = {
    "download.default_directory": download_path,  # Устанавливаем папку для загрузки
    "download.prompt_for_download": False,  # Отключаем запрос на подтверждение скачивания
    "directory_upgrade": True
}
chrome_options.add_experimental_option("prefs", prefs)

# Запуск браузера
driver = webdriver.Chrome(options=chrome_options)

# Открываем страницу с матчем
url = 'https://eternalesports.club/lobbies/3/19882b'  # Замени на актуальный URL
driver.get(url)

# Ждем, пока страница загрузится
time.sleep(3)

# Ищем кнопку для скачивания
download_button = driver.find_element(By.CSS_SELECTOR, "button.MuiButton-outlined")

# Прокручиваем страницу, чтобы кнопка стала видимой
driver.execute_script("arguments[0].scrollIntoView();", download_button)

# Увеличиваем время ожидания до 20 секунд и проверяем, что кнопка кликабельна
try:
    print("Waiting for button to become clickable...")
    wait = WebDriverWait(driver, 20)  # Увеличиваем таймаут до 20 секунд
    wait.until(EC.element_to_be_clickable((By.CSS_SELECTOR, "button.MuiButton-outlined")))

    # Пробуем кликнуть по кнопке
    download_button.click()
    print("Кнопка нажата!")
except Exception as e:
    print(f"Ошибка при клике по кнопке: {e}")
    # Если обычный клик не сработал, пытаемся кликнуть через JavaScript
    driver.execute_script("arguments[0].click();", download_button)
    print("Кнопка нажата через JavaScript!")

# Ждем завершения скачивания
time.sleep(3)

# Закрываем браузер
driver.quit()

print(f"Файл был скачан в папку: {download_path}")
